armory/data/datasets.py

```diff
@@ -20,70 +21,98 @@
 logger = logging.getLogger(__name__)
 
 
-def _in_memory_dataset_tfds(
-    dataset_name: str, dataset_dir: str, preprocessing_fn: Callable,
-) -> (np.ndarray, np.ndarray, np.ndarray, np.ndarray):
-    default_graph = tf.compat.v1.keras.backend.get_session().graph
+class ArmoryDataGenerator(DataGenerator):
+    def __init__(self, generator, size, batch_size, preprocessing_fn=None):
+        super().__init__(size, batch_size)
+        self.preprocessing_fn = preprocessing_fn
+        self.generator = generator
+        self.total_iterations = size // batch_size
 
-    train_x, train_y = tfds.load(
-        dataset_name,
-        batch_size=-1,
-        split="train",
-        as_supervised=True,
-        data_dir=dataset_dir,
-    )
+    def get_batch(self) -> (np.ndarray, np.ndarray):
+        x, y = next(self.generator)
+        if self.preprocessing_fn:
+            x = self.preprocessing_fn(x)
+
+        return x, y
+
+
+def _generator_from_tfds(
+    dataset_name: str,
+    split_type: str,
+    batch_size: int,
+    epochs: int,
+    dataset_dir: str,
+    preprocessing_fn: Callable,
+):
+    default_graph = tf.compat.v1.keras.backend.get_session().graph
 
-    test_x, test_y = tfds.load(
+    ds, ds_info = tfds.load(
         dataset_name,
-        batch_size=-1,
-        split="test",
+        batch_size=batch_size,
+        split=split_type,
         as_supervised=True,
         data_dir=dataset_dir,
+        with_info=True,
     )
 
-    train_x = tfds.as_numpy(train_x, graph=default_graph)
-    train_y = tfds.as_numpy(train_y, graph=default_graph)
-    test_x = tfds.as_numpy(test_x, graph=default_graph)
-    test_y = tfds.as_numpy(test_y, graph=default_graph)
+    ds = ds.repeat(epochs)
+    ds = ds.shuffle(batch_size * 10)
+    ds = ds.prefetch(tf.data.experimental.AUTOTUNE)
+    ds = tfds.as_numpy(ds, graph=default_graph)
 
-    if preprocessing_fn:
-        train_x = preprocessing_fn(train_x)
-        test_x = preprocessing_fn(test_x)
+    generator = ArmoryDataGenerator(
+        ds,
+        size=epochs * ds_info.splits[split_type].num_examples,
+        batch_size=batch_size,
+        preprocessing_fn=preprocessing_fn,
+    )
 
-    return train_x, train_y, test_x, test_y
+    return generator
 
 
 def mnist(
-    dataset_dir: str = None, preprocessing_fn: Callable = None,
-) -> (np.ndarray, np.ndarray, np.ndarray, np.ndarray):
+    split_type: str,
+    epochs: int,
+    batch_size: int,
+    dataset_dir: str = None,
+    preprocessing_fn: Callable = None,
+) -> ArmoryDataGenerator:
     """
     Handwritten digits dataset:
         http://yann.lecun.com/exdb/mnist/
-
-    returns:
-        train_x, train_y, test_x, test_y
     """
     if not dataset_dir:
         dataset_dir = paths.docker().dataset_dir
```

Comment:
Recommend moving this to the ArmoryDataGenerator class init instead of replicating these lines in every function.

Reply:
Done. Thanks!